### PR TITLE
[GraphQL] Option to collapse parts

### PR DESCRIPTION
**Pull request by Jetski5822**
At the moment when you do

```
{
	blog {
		autoroute {
			path
		}
	}
}
```

now if you go to the part on the type, and turn the option to collapse the part in, you will get

```
{
	blog {
		path
	}
}
```

### CODE REVIEW

**Comment by sebastienros** [comment]
I like the idea. I would have suggested to provide some kind of metadata that any module can provide to register fields specifically. The goal would be to keep the reflection based registration, but one could also decide to defined what fields are registered and how they are. This way we could decide that for AutoroutePart it's always better to expose it as `Path`. And if someone doesn't do anything then it's just the default reflection, `[partname].[property]`. We could even do it for fields, and for TextField we'd say that the name of the fields is `.Text` automatically. We would not do it for complex fields link Link, GeoLocation, ...

Also it feels weird to change how queries behave from configuration. I don't expect people would change it in the part settings if we have good defaults.

**Comment by Jetski5822** [comment]
Yeah, kind of agree, should probably be metadata... I’ll take a stab tomorrow at it.

Glad you like it

**Comment by Jetski5822** [comment]
Could do! No reason why not

**Comment by Jetski5822** [comment]
Okay, trying to remember the use-case an I think it goes like this...

We have two content types, `Author` and `Medicine Leaflet`, and we have a content part of "Person".

With the `Author` content type, we want to roll up the `Person` content part on to the type, but on the Medicine Leaflet we want it to remain so that on the `Medicine Leaflet` it looks like this

```
{
  medicineleaflet {
    author {
      name
    }
  }
}
```

but for Author it would be
```
{
  author {
    name
  }
}
```

How would I achieve this with metadata?

Hope that makes sense cc: / @carlwoodhouse

**Comment by carlwoodhouse** [comment]
Yeah it makes sense :)

**Comment by Jetski5822** [comment]
Im not sure whats left, @sebastienros really need your thoughts on this? Else can we merge it in please?

**Comment by carlwoodhouse** [comment]
ya really need this, or similar - with complex types / parts the ability to rollup and simplify is really needed! especially if you're wanting to expose the graphql api to non orchard consumers who have less idea about parts

**Comment by sebastienros** [comment]
We can compromise and keep the "collapse only" option you have done, but from metadata then, not directly from settings.

**Comment by Jetski5822** [comment]
Hmm how will that work on a per content type basis??

**Comment by Jetski5822** [comment]
```
@Jetski5822 Instead of storing the data in the settings as the source of thruth, create a class called "GraphQLMapping" that contains what you would have put in your settings. This way from my module I can resolve it and add something to it, instead of having only users change these from the admin. Then your settings can also be used to add more "mappings" to it. A mapping is what you call "this part is collapsed". I suggested we should have the mapping at the field level but if that is too much work I am fine with the part level, meaning all the fields of the part are then moved up (hoping there is no conflict with names then as you can't change the name of each field).
```

**Comment by Jetski5822** [comment]
Okay so I think this PR is ready, the rules are;

Collapsing works at on a heir-achy;
       
##Set in code
If the Content Type is marked as collapsed, then all parts that are contained are collapsed
If the Content Type is not marked as collapsed, then it falls down to the content part attached to the content type, has that been set as collapsed.
If the Content Part is marked as collapsed

##Set as settings in the UI attached to the Part on the Content Type
(If all of the above fail) The user then has the option to collapse

To set a part as collapsed by default, you do this;

```
services.SetPartAsCollapsed<AutoroutePart>();
```

By default I have set the AutoroutePart to be automatically collapsed by default. When collapsed in code, you cannot overwrite it in the UI.

UI for the collapse option on an autoroute part, which is disabled in the ui due to being set in code.

image

UI for the collapse option for a part where the option is available

image

By default, it will always be false.

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        /// <summary>
+        /// Collapsing works at a heirachy
+        ///
+        /// If the Content Type is marked at collapsed, then all parts are collapsed.
+        /// If the Content Type is not marked collapsed, then it falls down to the content type under it.
+        /// If the Content Part at a top level is marked collapsed, then it will trump above.
+        /// </summary>
+        /// <param name="definition"></param>
+        /// <returns></returns>
+        public bool ShouldCollapse(ContentTypePartDefinition definition)
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs]: The other way around. The module that adds the Settings editor will also update the options when the settings are changed, and on startup. The settings will "configure" the option. However using an Option looks weird, why not just a standard service class? Not sure which one is better.

Comment (Jetski5822) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs]: Options are fine IMO, I saw settings allowing the user to set it from the UI if not set in code.

**Pull request merged**